pkg/controllers/work/apply_controller.go

```diff
@@ -384,12 +394,55 @@ func (r *ApplyWorkReconciler) applyUnstructured(ctx context.Context, gvr schema.
 
 	// We only try to update the object if its spec hash value has changed.
 	if manifestObj.GetAnnotations()[manifestHashAnnotation] != curObj.GetAnnotations()[manifestHashAnnotation] {
+		// we need to merge the owner reference between the current and the manifest since we support one manifest
+		// belong to multiple work so it contains the union of all the appliedWork
+		manifestObj.SetOwnerReferences(mergeOwnerReference(curObj.GetOwnerReferences(), manifestObj.GetOwnerReferences()))
+		// record the raw manifest with the hash annotation in the manifest
+		if err := setModifiedConfigurationAnnotation(manifestObj); err != nil {
+			return nil, ManifestNoChangeAction, err
+		}
+		annotations := manifestObj.GetAnnotations()
+		if !isAnnotationsSizeValid(gvr, manifestRef, annotations) {
+			delete(annotations, lastAppliedConfigAnnotation)
+			manifestObj.SetAnnotations(annotations)
+			return r.applyObject(ctx, gvr, manifestObj)
+		}
 		return r.patchCurrentResource(ctx, gvr, manifestObj, curObj)
 	}
 
 	return curObj, ManifestNoChangeAction, nil
 }
 
+// isAnnotationSizeValid returns true if the total size of the annotations for a manifest is less than 256KB.
+func isAnnotationsSizeValid(gvr schema.GroupVersionResource, manifestRef klog.ObjectRef, annotations map[string]string) bool {
```

Comment:
Hi Arvind! For this method, maybe it is better idea to reuse the [ValidateAnnotationsSize](ValidateAnnotationsSize) function from the Kubernetes API validation package directly.